File: zig/basics/fibonacci.zig
// Print the first 20 Fibonacci numbers, carried in two mutable locals.
// Build: zig build-exe fibonacci.zig (written against Zig 0.16)

const std = @import("std");

pub fn main() void {
    var current: u64 = 0;
    var next: u64 = 1;
    var i: usize = 0;
    while (i < 20) : (i += 1) {
        if (i > 0) std.debug.print(" ", .{});
        std.debug.print("{d}", .{current});
        const following = current + next;
        current = next;
        next = following;
    }
    std.debug.print("\n", .{});
}

File: rust/basics/fibonacci.rs
// Print the first 20 Fibonacci numbers from a hand-written Iterator implementation.
// Build: rustc -o fibonacci fibonacci.rs

struct Fibonacci {
    current: u64,
    next: u64,
}

impl Iterator for Fibonacci {
    type Item = u64;

    fn next(&mut self) -> Option<u64> {
        let value = self.current;
        self.current = self.next;
        self.next += value;
        Some(value)
    }
}

fn main() {
    let fib = Fibonacci {
        current: 0,
        next: 1,
    };
    let values: Vec<String> = fib.take(20).map(|n| n.to_string()).collect();
    println!("{}", values.join(" "));
}
